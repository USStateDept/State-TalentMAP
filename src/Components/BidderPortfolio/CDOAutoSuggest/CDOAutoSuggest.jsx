import { Component } from 'react';
import PropTypes from 'prop-types';
import { isEmpty, isEqual, throttle } from 'lodash';
import { connect } from 'react-redux';
import Picky from 'react-picky';
import bowser from 'bowser';
import { EMPTY_FUNCTION } from 'Constants/PropTypes';
import { bidderPortfolioSelectCDOsToSearchBy } from 'actions/bidderPortfolio';
import { unsetClientView } from 'actions/clientView';
import ListItem from 'Components/BidderPortfolio/BidControls/BidCyclePicker/ListItem';
import filterUsers from '../helpers';

// TODO - Running into an issue where the label/span element is also
// passing up an event almost concurrently (400ms difference).
// This only happens in IE11, so we add a leading throttle so that
// any additional events within 1000ms do not get called.
const browser = bowser.getParser(window.navigator.userAgent);
const isIE = browser.satisfies({ 'internet explorer': '<=11' });
const THROTTLE_MS = isIE ? 1000 : 0;

export function renderList({ items, selected, ...rest }) {
  const getIsSelected = item => !!selected.find(f => f.id === item.id);
  return items.map(item => <ListItem key={item.id} item={item} {...rest} queryProp="name" getIsSelected={getIsSelected} />);
}

class CDOAutoSuggest extends Component {
  constructor(props) {
    super(props);
    this.selectMultipleOption = throttle(
      this.selectMultipleOption.bind(this),
      THROTTLE_MS,
      { trailing: false, leading: true },
    );
    this.state = {
      suggestions: filterUsers('', props.cdos),
    };
  }

  UNSAFE_componentWillReceiveProps(nextProps) {
    if (!isEqual(this.props.cdos, nextProps.cdos)) {
      this.setState({ suggestions: filterUsers('', nextProps.cdos) });
    }
    if (!isEqual(this.props.cdoPills, nextProps.cdoPills)) {
      this.selectMultipleOption(nextProps.cdoPills, true);
    }
  }

  selectMultipleOption(value, fromPills) {
    if (isEmpty(value) && fromPills && !isEmpty(this.props.currentCDO)) {
      this.props.setCDOsToSearchBy([this.props.currentCDO]);
    } else {
      this.props.setCDOsToSearchBy(value);
    }
  }

  render() {
    const { suggestions } = this.state;
    const { isLoading, hasErrored, selection, cdoSelections } = this.props; // eslint-disable-line
    return (
      !isLoading && !hasErrored &&
        <div className="cdo-autosuggest">
          <Picky
            placeholder="Select CDOs"
            value={selection}
            options={suggestions}
            onChange={this.selectMultipleOption}
            numberDisplayed={2}
            multiple
            includeFilter
            dropdownHeight={255}
            renderList={renderList}
            valueKey="id"
            labelKey="name"
            includeSelectAll

            // TODO - Once React 16 is integrated, upgrade to react-picky >=5.2.0 to use this prop
            // filterTermProcessor={trim} // import trim from lodash
          />
        </div>
    );
  }
}

CDOAutoSuggest.propTypes = {
  cdos: PropTypes.arrayOf(PropTypes.shape({})),
  isLoading: PropTypes.bool,
  hasErrored: PropTypes.bool,
  selection: PropTypes.arrayOf(PropTypes.shape({})),
  setCDOsToSearchBy: PropTypes.func,
  cdoPills: PropTypes.arrayOf(PropTypes.shape({})),
  currentCDO: PropTypes.shape({}),
};

CDOAutoSuggest.defaultProps = {
  cdos: [],
  isLoading: false,
  hasErrored: false,
  selection: [],
  setCDOsToSearchBy: EMPTY_FUNCTION,
  cdoPills: [],
  currentCDO: {},
};

const mapStateToProps = state => ({
  cdos: state.bidderPortfolioCDOs,
  isLoading: state.bidderPortfolioCDOsIsLoading,
  hasErrored: state.bidderPortfolioCDOsHasErrored,
  selection: state.bidderPortfolioSelectedCDOsToSearchBy,
  currentCDO: state.bidderPortfolioSelectedCDO,
});

export const mapDispatchToProps = dispatch => ({
  setCDOsToSearchBy: (arr) => {
    dispatch(bidderPortfolioSelectCDOsToSearchBy(arr));
    dispatch(unsetClientView());
  },
});

export default connect(mapStateToProps, mapDispatchToProps)(CDOAutoSuggest);
