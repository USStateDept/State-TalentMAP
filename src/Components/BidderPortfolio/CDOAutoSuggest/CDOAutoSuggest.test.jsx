import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import sinon from 'sinon';
import CDOAutoSuggest, { renderList } from './CDOAutoSuggest';

describe('renderList', () => {
  it('render', () => {
    const wrapper = renderList({ items: [{ id: 1 }, { id: 2 }], selected: [{ id: 1 }] });
    expect(wrapper).toBeDefined();
  });
});

describe('CDOAutoSuggest', () => {
  const props = {
    cdos: [
      { first_name: 'John', last_name: 'Daniels', id: 3, hru_id: 3 },
      { first_name: 'Mary', last_name: 'Brown', id: 4, hru_ud: 4 },
    ],
    selection: [{ first_name: 'Mary', last_name: 'Brown', id: 4 }],
    isLoading: false,
    hasErrored: false,
    setCDOsToSearchBy: () => {},
  };

  it('is defined', () => {
    const wrapper = shallow(<CDOAutoSuggest.WrappedComponent {...props} />);
    expect(wrapper).toBeDefined();
  });

  it('is defined after new props', () => {
    const wrapper = shallow(<CDOAutoSuggest.WrappedComponent {...props} />);
    wrapper.setProps(props);
    expect(wrapper).toBeDefined();
  });

  it('calls the prop on onSuggestionSelected()', () => {
    const spy = sinon.spy();
    const wrapper = shallow(<CDOAutoSuggest.WrappedComponent {...props} setCDOsToSearchBy={spy} />);
    wrapper.instance().selectMultipleOption([props.cdos[0]]);
    sinon.assert.calledOnce(spy);
  });

  it('matches snapshot', () => {
    const wrapper = shallow(<CDOAutoSuggest.WrappedComponent {...props} />);
    expect(toJSON(wrapper)).toMatchSnapshot();
  });
});
