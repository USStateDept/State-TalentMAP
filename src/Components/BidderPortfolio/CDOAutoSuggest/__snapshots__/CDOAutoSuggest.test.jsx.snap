// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CDOAutoSuggest matches snapshot 1`] = `
<div
  className="cdo-autosuggest"
>
  <r
    dropdownHeight={255}
    filterDebounce={150}
    id="picky"
    includeFilter={true}
    includeSelectAll={true}
    keepOpen={true}
    labelKey="name"
    multiple={true}
    numberDisplayed={2}
    onChange={[Function]}
    options={
      Array [
        Object {
          "first_name": "Mary",
          "hru_ud": 4,
          "id": 4,
          "last_name": "Brown",
        },
        Object {
          "first_name": "John",
          "hru_id": 3,
          "id": 3,
          "last_name": "Daniels",
        },
      ]
    }
    placeholder="Select CDOs"
    renderList={[Function]}
    selectAllMode="default"
    selectAllText="Select all"
    tabIndex={0}
    value={
      Array [
        Object {
          "first_name": "Mary",
          "id": 4,
          "last_name": "Brown",
        },
      ]
    }
    valueKey="id"
  />
</div>
`;
